src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -693,3 +697,79 @@ g.test('empty_bind_group_layouts_requires_empty_bind_groups')
       encoder.finish();
     }, !_success);
   });
+
+g.test('empty_bind_group_layouts_requires_empty_bind_groups,render_pass')
+  .desc(
+    `
+  Test that a render pipeline with empty bind groups layouts requires empty bind groups to be set.
+  `
+  )
+  .paramsSimple([
+    { bindGroupLayoutEntryCount: 4, _success: true }, // Control case
+    { bindGroupLayoutEntryCount: 3, _success: false },
+  ])
+  .fn(async t => {
+    const { bindGroupLayoutEntryCount, _success } = t.params;
+
+    const emptyBGL = t.device.createBindGroupLayout({ entries: [] });
+    const emptyBGLs = [];
+    for (let i = 0; i < 4; i++) {
+      emptyBGLs.push(emptyBGL);
+    }
+
+    const pipelineLayout = t.device.createPipelineLayout({
+      bindGroupLayouts: emptyBGLs,
+    });
+
+    const colorFormat = 'rgba8unorm';
+    const pipeline = t.device.createRenderPipeline({
+      layout: pipelineLayout,
+      vertex: {
+        module: t.device.createShaderModule({
+          code: `@vertex fn main() -> @builtin(position) vec4<f32> { return vec4<f32>(); }`,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: t.device.createShaderModule({
+          code: `@fragment fn main() {}`,
+        }),
+        entryPoint: 'main',
+        targets: [{ format: colorFormat, writeMask: 0 }],
+      },
+    });
+
+    const emptyBindGroup = t.device.createBindGroup({
+      layout: emptyBGL,
+      entries: [],
+    });
+
+    const encoder = t.device.createCommandEncoder();
+
+    const attachmentTexture = t.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 16, height: 16, depthOrArrayLayers: 1 },
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+    });
+
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: attachmentTexture.createView(),
+          clearValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+    });
+
+    renderPass.setPipeline(pipeline);
+    for (let i = 0; i < bindGroupLayoutEntryCount; i++) {
+      renderPass.setBindGroup(i, emptyBindGroup);
+    }
```

Comment:
The draw call would go here.

Ideally, we should test all 4 draw calls because we want to make sure that all 4 do the validation. See `api/validation/encoding/cmds/render/draw.spec.ts` and `api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts` for examples of tests where we did parameterization like this. (I searched for `'drawIndexedIndirect'`, with the quotation marks, to find these, because I guessed that was the wording we used in the case parameters).

Reply:
Oh, I just realized that this *is* `pipeline_bind_group_compat.spec.ts`. So I think you can use the `doRender` helper here and the `doCompute` helper in the compute test.

Reply:
Thank you for adding TODOs for doRender/doCompute. Let me fix them in the following PR.